cps/kobo.py

```diff
@@ -0,0 +1,523 @@
+            },
+        }
+        return b64encode_json(token)
+
+
+@kobo.route("/v1/library/sync")
+@login_required
+@download_required
+def HandleSyncRequest():
+    sync_token = SyncToken.from_headers(request.headers)
+    log.info("Kobo library sync request received.")
+
+    # TODO: Limit the number of books return per sync call, and rely on the sync-continuatation header
+    # instead so that the device triggers another sync.
+
+    new_books_last_modified = sync_token.books_last_modified
+    new_books_last_created = sync_token.books_last_created
+    entitlements = []
+
+    # sqlite gives unexpected results when performing the last_modified comparison without the datetime cast.
+    # It looks like it's treating the db.Books.last_modified field as a string and may fail
+    # the comparison because of the +00:00 suffix.
+    changed_entries = (
+        db.session.query(db.Books)
+        .filter(func.datetime(db.Books.last_modified) > sync_token.books_last_modified)
+        .all()
+    )
+    for book in changed_entries:
+        entitlement = {
+            "BookEntitlement": create_book_entitlement(book),
+            "BookMetadata": get_metadata(book),
+            "ReadingState": reading_state(book),
+        }
+        if book.timestamp > sync_token.books_last_created:
+            entitlements.append({"NewEntitlement": entitlement})
+        else:
+            entitlements.append({"ChangedEntitlement": entitlement})
+
+        new_books_last_modified = max(
+            book.last_modified, sync_token.books_last_modified
+        )
+        new_books_last_created = max(book.timestamp, sync_token.books_last_modified)
+
+    sync_token.books_last_created = new_books_last_created
+    sync_token.books_last_modified = new_books_last_modified
+
+    # Missing feature: Detect server-side book deletions.
+
+    # Missing feature: Join the response with results from the official Kobo store so that users can still buy and access books from the device store (particularly while on-the-road).
+
+    response = make_response(jsonify(entitlements))
+
+    sync_token.to_headers(response.headers)
+    response.headers["x-kobo-sync-mode"] = "delta"
+    response.headers["x-kobo-apitoken"] = "e30="
+    return response
+
+
+@kobo.route("/v1/library/<book_uuid>/metadata")
+@login_required
+@download_required
+def HandleMetadataRequest(book_uuid):
+    log.info("Kobo library metadata request received for book %s" % book_uuid)
+    book = db.session.query(db.Books).filter(db.Books.uuid == book_uuid).first()
+    if not book or not book.data:
+        log.info(u"Book %s not found in database", book_uuid)
+        return make_response("Book not found in database.", 404)
+
+    metadata = get_metadata(book)
+    return jsonify([metadata])
+
+
+def get_download_url_for_book(book, book_format):
+    return "{url_base}/download/{book_id}/{book_format}".format(
+        url_base=config.config_server_url,
```

Comment:
I could solve the server name issue by using:

request.environ['werkzeug.request'].base_url 

This worked on with apache as reverse proxy and with direct connection and removes the new config option

Reply:
I also discovered Flask.url_for so I've updated all usages I added to that and reverted the config_server_url changes!

Reply:
I also discovered Flask.url_for so I've updated all usages I added to that and reverted the config_server_url changes!